### PR TITLE
Fix a small issue in sigma library

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

**Comment by nopslide** [comment]
This pull request **introduces 3 alerts** when merging b1fc991c4533e51abff3afae5eab5ea607c9c7ef into 69bb948f6836f3d5c9772326880c390488534166 - [view on LGTM.com](https://lgtm.com/projects/g/zcoinofficial/zcoin/rev/pr-d1d80356111c030cf523fc96fe5063f0dc8fde5b)

**new alerts:**

* 3 for &#39;import \*&#39; may pollute namespace

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by ultimaweapon** [comment]
@martun could you merge latest `sigma` into this branch first so we will only see a new changes that just introduced by this PR?

File: src/sigma/sigmaplus_prover.hpp [hunk 1]
```diff
@@ -18,27 +18,30 @@ void SigmaPlusProver<Exponent, GroupElement>::proof(
         std::size_t l,
         const Exponent& r,
         SigmaPlusProof<Exponent, GroupElement>& proof_out) {
-    int N = commits.size();
+    std::size_t N = commits.size();
```
Comment (ultimaweapon) [inline src/sigma/sigmaplus_prover.hpp]: Nice one for make it proper.

File: src/sigma/sigma_primitives.hpp [hunk 2]
```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
...
+void SigmaPrimitives<Exponent, GroupElement>::generate_challenge(
+        const std::vector<GroupElement>& group_elements,
         Exponent& result_out) {
     CSHA256 hash;
-    unsigned char data[3 * A.serialize_size];
-    unsigned char* current = A.serialize(data);
-    current = C.serialize(current);
-    D.serialize(current);
-    hash.Write(data, 3 * A.memoryRequired());
+    std::vector<unsigned char> data(group_elements.size() * group_elements[0].memoryRequired());
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: This possible to crash if `group_elements` is empty.

File: src/sigma/sigma_primitives.hpp [hunk 3]
```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
...
+    unsigned char* current = data.data();
+    for (size_t i = 0; i < group_elements.size(); ++i) {
+        current = group_elements[i].serialize(current);
+    }
+    hash.Write(data.data(), data.size());
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: IIRC `hash.Write` will calculate a hash of data based of previous write. That mean we don't need to concatenate into `data`  by ourself.

File: src/sigma/sigma_primitives.hpp [hunk 4]
```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
...
     hash.Finalize(result_data);
     result_out = result_data;
 }
 
 template<class Exponent, class GroupElement>
 void SigmaPrimitives<Exponent, GroupElement>::new_factor(
-        Exponent x,
-        Exponent a,
+        const Exponent& x,
+        const Exponent& a,
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: Nice once for improvement!

File: src/sigma/sigma_primitives.hpp [hunk 5]
```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
...
         std::vector<Exponent>& coefficients) {
     std::vector<Exponent> temp;
     temp.resize(coefficients.size() + 1);
     for (std::size_t j = 0; j < coefficients.size(); j++)
-        temp[j] += x * coefficients[j];
+        temp[j] = x * coefficients[j];
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: Nice one for fixing weird looking code.

File: src/sigma/sigma_primitives.hpp [hunk 6]
```diff
@@ -71,34 +63,33 @@ std::vector<uint64_t> SigmaPrimitives<Exponent, GroupElement>::convert_to_nal(
...
     for(std::size_t j = 0; j < coefficients.size(); j++)
         temp[j + 1] += a * coefficients[j];
-    coefficients = temp;
+    coefficients = std::move(temp);
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: Good improvement!

File: src/sigma/r1_proof_verifier.hpp [hunk 7]
```diff
@@ -16,61 +16,77 @@ R1ProofVerifier<Exponent,GroupElement>::R1ProofVerifier(
...
+
+template<class Exponent, class GroupElement>
+bool R1ProofVerifier<Exponent,GroupElement>::verify_final_response(
+            const R1Proof<Exponent, GroupElement>& proof,
+            const Exponent& challenge_x,
+            std::vector<Exponent>& f_out) const {
+    const std::vector<Exponent>& f = proof.f_;
+    f_out.reserve(n_ * m_);
+    for(int j = 0; j < m_; ++j) {
+        f_out.push_back(Exponent(uint64_t(0)));
```
Comment (ultimaweapon) [inline src/sigma/r1_proof_verifier.hpp]: Don't know it is intention or not to not clear previous items in `f_out`. If it not intention we should clear it before adding new items.

File: src/sigma/r1_proof_generator.hpp [hunk 8]
```diff
@@ -13,71 +13,90 @@ R1ProofGenerator<Exponent,GroupElement>::R1ProofGenerator(
...
+        R1Proof<Exponent, GroupElement>& proof_out) {
     //f
-    std::vector<Exponent> f;
-    f.reserve(m_ * (n_ - 1));
-    for(int j = 0; j < m_; j++){
+    proof_out.f_.reserve(m_ * (n_ - 1));
+    for(int j = 0; j < m_; j++) {
         for(int i = 1; i < n_; i++)
-        f.emplace_back(b_[(j * n_) + i] * x + a[(j * n_) + i]);
+        proof_out.f_.emplace_back(b_[(j * n_) + i] * challenge_x + a[(j * n_) + i]);
```
Comment (ultimaweapon) [inline src/sigma/r1_proof_generator.hpp]: Same as previous issue. If it not intension clear `proof_out` before adding new one will be more safe.

File: src/sigma/sigma_primitives.hpp (hunk 2 above)
Comment (martun) [reply src/sigma/sigma_primitives.hpp]: It can not be empty. Added a small check.

File: src/sigma/r1_proof_verifier.hpp (hunk 7 above)
Comment (martun) [reply src/sigma/r1_proof_verifier.hpp]: We expect it to be empty, so added a line to clear it anyway.

File: src/sigma/r1_proof_generator.hpp (hunk 8 above)
Comment (martun) [reply src/sigma/r1_proof_generator.hpp]: done

**Comment by martun** [comment]
I deleted 1 unit test because it was not necessary any more. It was mimicking some part of the sigma proof. It was there for debugging purposes, but not needed any more. There are other unit tests which cover all functionality.

File: src/sigma/sigma_primitives.hpp [hunk 9]
```diff
@@ -67,7 +67,7 @@ void SigmaPrimitives<Exponent, GroupElement>::generate_challenge(
         const std::vector<GroupElement>& group_elements,
         Exponent& result_out) {
     if (group_elements.empty())
-        throw ZerocoinException("Group elements empty while generating a challenge.");
+        throw std::runtime_error("Group elements empty while generating a challenge.");
```
Comment (ultimaweapon) [inline src/sigma/sigma_primitives.hpp]: Will be better to use `std::invalid_argument` instead. But can be fixed later.

**Comment by ultimaweapon** [comment]
Still need someone who understand mathematics to review business logic. I approve only programming issues.

**Pull request merged**